app/models/mixins/retirement_mixin.rb

```diff
@@ -231,15 +231,35 @@ def valid_zone?
     respond_to?(:my_zone) && my_zone.present?
   end
 
+  def system_context_requester
+    if try(:evm_owner_id).present?
+      User.find(evm_owner_id)
+    else
+      $log.info("System context defaulting to admin user because owner of #{name} not set.")
+      User.find_by(:userid => 'admin')
+    end
+  end
+
+  def q_user_info(q_options, requester)
+    if requester.present?
+      if requester.kind_of?(String)
+        requester = User.find_by(:userid => requester)
+      end
+      q_options[:user_id] = requester.id
+      if requester.current_group.present? && requester.current_tenant.present?
+        q_options[:group_id] = requester.current_group.id
+        q_options[:tenant_id] = requester.current_tenant.id
+      end
+    end
+    q_options
+  end
+
   def setup_event_hash(requester)
-    event_hash = {:retirement_initiator => "system"}
-    event_hash[retirement_base_model_name.underscore.to_sym] = self
-    event_hash[:host] = host if self.respond_to?(:host)
-    if requester
-      event_hash[:userid] = requester
-      event_hash[:retirement_initiator] = "user"
+    {}.tap do |event|
+      event[:userid] = requester
+      event[retirement_base_model_name.underscore.to_sym] = self
+      event[:host] = host if respond_to?(:host)
+      event[:type] ||= self.class.name
```

Comment:
Why the `||`?

Reply:
Because Madhu. https://github.com/ManageIQ/manageiq/commit/e30a757759bc539fec42fcc13023cfe1ccaad29d#diff-5511f6e21f2abb2c6c79f6cc1ac1d6a9R201

Reply:
Good job using `tap`, but there is room for improvement.  I would expect that the keys you are always setting are setup in the initial hash and conditional keys would be addressed in the `tap` section.

Something like this:
```ruby
{
  :userid => requester,
  retirement_base_model_name.underscore.to_sym => self,
  :type => self.class.name
}.tap do |event|
  event[:host] = host if respond_to?(:host)
end
```